feat: ✨ decision post on why to use git-cliff for changelogs (#273)
seedcase-project/decisions: why-changelog-with-git-cliff/index.qmd

---
title: "Why git-cliff for changelog generation"
description: "Automatically generating a changelog can save a lot of administrative time, especially for a small group like ours. This post explains our reasons for using git-cliff to auto-generate changelogs."
date: "2026-05-22"
categories:
  - develop
  - git
---

## Context and problem statement

::: content-hidden
State the context and some background on the issue, then write a statement in
the form of a question for the problem.
:::

We originally used Commitizen for changelog generation, but we encountered some
issues and limitations to it as our needs grew. So this post is explaining why
we switched to git-cliff. See our [previous post on using Commitizen for
changelog generation](/why-changelog-with-commitizen/index.qmd) for more

> information about our context and problem statement as well as the original

Reviewer: ```suggestion
information about the context and problem statement as well as the original
```

reasons for using Commitizen.

> This post explains some of our issues and why we switched to something else.

Reviewer: I think it would make sense to restate/summarise the context and problem statement here.

> This post explains some of our issues and why we switched to something else.

Reviewer: ```suggestion
This post explains some of our issues with Commitizen and why we switched to git-cliff.
```

::: content-hidden
List some reasons for why we need to make this decision and what things have
arisen that impact work.
:::

Commitizen worked fairly well for our needs for a while, but as we incorporated
it in other, non-Python projects, we quickly started encountering some issues
and limitations. Some of the issues we encountered include:

- The changelog generated by Commitizen is not as easy to customise. For
  example, in R packages, the website generator
  [pkgdown](https://pkgdown.r-lib.org/) expects a specific format for the
  `NEWS.md` file, but we couldn't (easily) modify the changelog template to meet
  these requirements.
- When generating the changelog to be included in the GitHub release notes, it
  would often add a commit message at the end of the changelog. It wasn't clear
  how to remove it or even where it came from, as the CLI didn't produce the
  same output.
- We wanted to be able to customise the changelog for different projects, like
  the previously mentioned R packages, but also for documentation projects,
  which include a lot more changes outside of `refactor`, `fix`, and `feat`. So
  some relevant changes to those repos would often be missing from the

>   changelog.

Reviewer: ```suggestion
- The changelog generated by Commitizen is not as easy to customise, but we need to be able to customise the changelog for different projects, like
  the Python packages, R packages, and documentation projects. For example, the website generator
  [pkgdown](https://pkgdown.r-lib.org/) for R packages expects a specific format for the
  `NEWS.md` file, but we couldn't (easily) modify the changelog template to meet
  these requirements. Documentation projects include a lot more changes outside of `refactor`, `fix`, and `feat`, and therefore
  some relevant changes to those repos would often be missing from the
  changelog.
- When generating the changelog to be included in the GitHub release notes, it
  would often add a commit message at the end of the changelog. It wasn't clear
  how to remove it or even where it came from, as the CLI didn't produce the
  same output.
```
Seems like item 1 and 3 are very closely related, so attempted to merge then.

- We wanted something built in a more "formal" programming language, especially
  with the increase in supply chain attacks on e.g. Node.js packages. This isn't

>   critical, but it does need to support diverse project types, not just e.g.

Reviewer: not critical for us bc we don't work in js, or?

>   critical, but it does need to support diverse project types, not just e.g.

Author: ```suggestion
  a critical need for us, but it does need to support diverse project types, not just e.g.
```

  Python projects.

## Considered options

::: content-hidden
List and describe some of the options, as well as some of the benefits and
drawbacks for each option.
:::

There are numerous tools available for automatically generating changelogs for
Git repositories following the Conventional Commit specification, as seen by
[this
list](https://www.conventionalcommits.org/en/about/#tooling-for-conventional-commits)

> by Conventional Commits. We have considered the following options:

Reviewer: ```suggestion
by Conventional Commits. We considered the following options:
```

- [Cocogitto](https://cocogitto.github.io/)
- [git-cliff](https://git-cliff.org/docs/)

> These were the only two that fit our needs and importantly were still actively

Reviewer: ```suggestion
These were the only two that fit our needs for customisation and security that is still actively
```
"security" as described in L51-53

maintained. Most of the other options hadn't been updated in years.

### Cocogitto

[Cocogitto](https://cocogitto.github.io/) is primarily focused on bumping
versions and helping to release them, checking commits match Conventional
Commits style, and for creating Conventional Commits. However, it does have a
changelog generation feature.

::: columns
::::: column
#### Benefits

- Is very well developed and documented.
- Can create custom changelogs using the [tera](https://keats.github.io/tera/)
  templating engine.
- Written in Rust, which is a powerful, statically-typed language that is known
  for its performance and safety.
:::::

::::: column
#### Drawbacks

> - Not nearly as easily customisable as git-cliff (see below).

Reviewer: ```suggestion
- Not as easily customisable as git-cliff (see below).
```

:::::
:::

### git-cliff

[git-cliff](https://git-cliff.org/docs/) is a changelog generator that can
generate changelogs for Git repositories that follow the Conventional Commits
specification. It can be customised to generate changelogs in different formats
with different filters.

::: columns
::::: column
#### Benefits

- Can generate changelogs based on commit messages following the SemVer and
  Conventional Commits specification.

> - Is very well-documented.

Reviewer: ```suggestion
- Is very well documented.
```

- Is actively maintained.
- Can generate changelogs in different formats.
- Allows for filtering and pre-processing of commits based on different criteria
  and patterns, which is really useful for us since we have different types of
  projects that require different processing needs.

> - Also written in Rust.

Reviewer: ```suggestion
- Is written in Rust.
```

> - It's output is consistent and predictable, which means we can incorporate it

Reviewer: ```suggestion
- Produces consistent and predictable output, which means we can incorporate it
```

  easily into our workflow for generating GitHub release notes.

> - It can be easily installed and used through `uv` via `uvx`.

Reviewer: ```suggestion
- Can be easily installed and used through `uv` via `uvx`.
```

:::::

::::: column
#### Drawbacks

> - It is primarily focused on changelog generation, so we would have to use

Reviewer: ```suggestion
- Is primarily focused on changelog generation (e.g. can bump versions, but isn't as easy to do), so we would have to use
```

Reviewer: primarily? What's other focus areas for them?

  another tool for linting and version bumping.
:::::
:::

## Decision outcome

We have decided to use git-cliff because it is highly customisable, which means
we can make it fit our needs for all our different projects.

### Consequences

- Because of our need to make custom changelogs for different projects, the
  maintenance cost will be higher than if we used something with defaults that
  were closer to what we needed. But unfortunately, there isn't a tool that can
  meet all our needs without some customisation, so we have to accept this
  trade-off.
- We introduce a new dependency, which brings with it some costs to security and
  maintenance. But git-cliff is actively maintained and written in Rust (which

>   is known for its safety), so we accept this trade-off as well.

Reviewer: Maybe there's also a consequence to be included here related to the drawback; we need to use another tool for linting and version bumping?
